### PR TITLE
Deprecate layer.position and layer.coordinates

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR supersedes #2249 and will deprecate `layer.position` and `layer.coordinates`. It adds `event.position` to the mouse events, which is the position of the cursor in world coordinates. It also removes some of the deprecated ability to call `layer.get_value()` and `layer.get_status()` without any position and the `layer.status` attribute. It also makes the `layer.world_to_data` transform public, and to get the old `layer.coordinates` you can do `layer.world_to_data(event.position)` or `layer.world_to_data(viewer.cursor.position)` depending on where you are/ what you want.

## Type of change

- [x] Deprecation

### CODE REVIEW

**Comment by sofroniewn** [comment]
All tests have been fixed and this is now ready for review @jni @tlambert03. The PR is fully backwards compatible apart from things that were already deprecated and flagged to be removed this release.

File: napari/_qt/qt_viewer.py
Comment (tlambert03) [inline napari/_qt/qt_viewer.py:712]: nice

File: napari/_tests/test_with_screenshot.py
Comment (tlambert03) [inline napari/_tests/test_with_screenshot.py:377]: it's fine if you don't want to deal with it in this PR... but I don't love that we create a new named tuple that behaves like an event for testing here.  seems like we might as well just create an actual `utils.events.Event`.

File: napari/_qt/qt_viewer.py
Comment (tlambert03) [inline napari/_qt/qt_viewer.py:677]: this is also peripheral to this PR: but I'm not convinced we need to be wrapping all of these events with the `ReadOnlyWrapper`.  It kinda feels overly-cautious, and adds object creation overhead to a _very_ frequent thing.

File: napari/_qt/qt_viewer.py
Comment (tlambert03) [inline napari/_qt/qt_viewer.py:674]: mentioned in summary, can this be `event.pos_world`?

**Review by tlambert03 — APPROVED** [review]
this looks good to me @sofroniewn!  couple thoughts I had as I looked through it:

- can we use an attribute name that is a bit more explicit since now the event object will have both `event.pos` (canvas) and `event.position` (world).  maybe call it `event.pos_world`? 
- perhaps for another PR, I feel like the need for ReadOnlyWrapper is questionable, particularly for an extremely frequent thing.

I'm approving, since none of those are particularly critical, but let me know if you make changes and want another review

File: napari/_tests/test_with_screenshot.py
Comment (sofroniewn) [reply napari/_tests/test_with_screenshot.py:377]: Yup 100% agreed, I'd probably rather do in another PR. I will make issue before merge

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py:677]: Yeah that's fair, though I also don't mind it as the event does get persisted through multiple callbacks. To be honest I think most events should really be read only by default, but maybe that's a discussion for another time. Right now thought it's actually not possible to just drop because of how persistence is done through the generator, so it would require more work. Happy to flag in a followup issue.

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py:674]: I see how this is confusing having both `event.pos` given by vispy which is position in canvas coordinates and `event.position` which is now our position in world coordinates, but I like the similarity between `viewer.cursor.position` and `event.position` which now are the same. I'd much rather rename `event.pos` to `event.canvas_position` or `event.position_canvas` at which point having `event.position` I think is ok. I also didn't and don't plan to put the `data_position` on the event (I had explored this in other PRs, but don't think it is needed).

My preference would be to stick with `event.position` for this property, which I think will be the most commonly used one, and then gradually rename some of the vispy ones as we make our own event here in followup PRs.

**Comment by sofroniewn** [comment]
> I'm approving, since none of those are particularly critical, but let me know if you make changes and want another review

Thanks for the review @tlambert03, I've gone through above and below and given detailed responses to the three key comments. I basically am either unsure or would prefer not to make any of those changes in this PR, but will open an issue with all three of them that could potentially be addressed after merge. They related much more to the mouse event/ mouse handling then deprecating of layer.position and layer.coordinates.

I will wait for an additional opinion from @jni before merge though, particularly on the attribute naming question which is probably the most important.

-------------

> can we use an attribute name that is a bit more explicit since now the event object will have both event.pos (canvas) and event.position (world). maybe call it event.pos_world?

As noted in detailed comment above I like `event.position` and would prefer to rename `event.pos` and any other confusing vispy attributes when we turn the mouse event into our own event

> perhaps for another PR, I feel like the need for ReadOnlyWrapper is questionable, particularly for an extremely frequent thing.

I potentially agree, but it's a little complicated to do right now due to all the generator stuff so it will need to be it's own PR

> I don't love that we create a new named tuple that behaves like an event for testing here. seems like we might as well just create an actual utils.events.Event

Agreed, I'll add this to an issue.

File: napari/_qt/qt_viewer.py
Comment (tlambert03) [reply napari/_qt/qt_viewer.py:674]: I'd probably vote for event.pos_canvas, and event.pos_world and _neither_ event.pos or event.position.   But I'm definitely ok with whatever you and @jni ultimately choose here.

**Comment by tlambert03** [comment]
> I basically am either unsure or would prefer not to make any of those changes in this PR,

Sounds good!  They were all minor.

File: napari/_qt/qt_viewer.py
Comment (jni) [reply napari/_qt/qt_viewer.py:674]: Oof. Tough one. I love brevity but I do think it's good to be clear here on which position is which. I *still I do like @tlambert03's suggestion of specifying *both*, `event.pos_canvas` and `event.pos_world` or `event.position_canvas` and `event.position_world`. What do you think of that suggestion @sofroniewn?

fwiw I'm not sure I agree that world coordinates will be the most common use — wouldn't you think that many people will want to access the data based on the cursor?

Comment (jni) [reply napari/_qt/qt_viewer.py:674]: oops, I just realised canvas != data. Gonna be a long week LOL

Comment (jni) [reply napari/_qt/qt_viewer.py:674]: ok, given the above, I'm happy with just `event.position` for now, with the plan to rename `.pos` to `.position_canvas`.

File: napari/_qt/qt_viewer.py
Comment (jni) [reply napari/_qt/qt_viewer.py:677]: > because of how persistence is done through the generator

can you elaborate?

I agree with @tlambert03's concerns though, this happens on each mouse move, it'd be at least interesting to measure how long creating the ROW takes.

File: napari/layers/base/base.py
Comment (jni) [inline napari/layers/base/base.py:942]: Very nice message. 😊

File: napari/layers/labels/_labels_mouse_bindings.py
Comment (jni) [inline napari/layers/labels/_labels_mouse_bindings.py:24]: 👍!

**Review by jni — APPROVED** [review]

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py:674]: Thanks for all the input here, I will leave as is for now, but will flag in issue as something to do

**Comment by sofroniewn** [comment]
Issue made in #2459, I will merge now

File: napari/_qt/qt_viewer.py
Comment (sofroniewn) [reply napari/_qt/qt_viewer.py:677]: It takes around 3us which isn't too bad, given at 60Hz frame rate is ~16.6ms.

The event gets "saved" here https://github.com/napari/napari/blob/c2b254f9b7a54f1256f0bb853ec7875c9deea4bd/napari/utils/interactions.py#L145-L146, it's a little complex. We should probably simplify.

**Pull request merged**
